internal/transform/contract_events.go

```diff
@@ -0,0 +1,144 @@
+package transform
+
+import (
+	"encoding/base64"
+	"fmt"
+
+	"github.com/stellar/stellar-etl/internal/toid"
+	"github.com/stellar/stellar-etl/internal/utils"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/strkey"
+	"github.com/stellar/go/xdr"
+)
+
+// TransformContractEvent converts a transaction's contract events and diagnostic events from the history archive ingestion system into a form suitable for BigQuery.
+// It is known that contract events are a subset of the diagnostic events XDR definition. We are opting to call all of these events
+// contract events for better clarity to data analytics users.
+func TransformContractEvent(transaction ingest.LedgerTransaction, lhe xdr.LedgerHeaderHistoryEntry) ([]ContractEventOutput, error) {
+	ledgerHeader := lhe.Header
+	outputTransactionHash := utils.HashToHexString(transaction.Result.TransactionHash)
+	outputLedgerSequence := uint32(ledgerHeader.LedgerSeq)
+
+	transactionIndex := uint32(transaction.Index)
+
+	outputTransactionID := toid.New(int32(outputLedgerSequence), int32(transactionIndex), 0).ToInt64()
+
+	outputCloseTime, err := utils.TimePointToUTCTimeStamp(ledgerHeader.ScpValue.CloseTime)
+	if err != nil {
+		return []ContractEventOutput{}, fmt.Errorf("for ledger %d; transaction %d (transaction id=%d): %v", outputLedgerSequence, transactionIndex, outputTransactionID, err)
+	}
+
+	// GetDiagnosticEvents will return all contract events and diagnostic events emitted
+	contractEvents, err := transaction.GetDiagnosticEvents()
+	if err != nil {
+		return []ContractEventOutput{}, err
+	}
+
+	var transformedContractEvents []ContractEventOutput
+
+	for _, contractEvent := range contractEvents {
+		var outputContractId string
+
+		outputInSuccessfulContractCall := contractEvent.InSuccessfulContractCall
+		event := contractEvent.Event
+		outputType := event.Type
+		outputTypeString := event.Type.String()
+
+		eventTopics := getEventTopics(event.Body)
+		outputTopics, outputTopicsDecoded := serializeScValArray(eventTopics)
+
+		eventData := getEventData(event.Body)
+		outputData, outputDataDecoded := serializeScVal(eventData)
+
+		// Convert the xdrContactId to string
+		// TODO: https://stellarorg.atlassian.net/browse/HUBBLE-386 this should be a stellar/go/xdr function
+		if event.ContractId != nil {
```

Comment:
In theory it can be. In practice I'm not sure.
I'm not sure there is an alternative here though. `ContractEvent.ContractId` is a pointer so it could be null at any point. But I think your point is does a contract/diagnostic event ever make sense if there is no ContractId.

Reply:
yeah, sorry I should have been more explicit. Should we be throwing an error here if the contractid is nil? Sounds like we may not know? 🤷‍♀️

Reply:
Yeah I'm not sure. I think for now we can make a dbt test/scheduled query to alert us if null contract_ids we above a certain threshold. If a large percentage of events are missing the contract_id we can investigate cause for all we know it might be attached to `config_settings`or other valid non-contract events

Reply:
yeah, sounds good